compiler/backend/src/org/jetbrains/kotlin/codegen/inline/MaxStackFrameSizeAndLocalsCalculator.java

```diff
@@ -336,10 +337,10 @@ public void visitMaxs(int maxStack, int maxLocals) {
          */
         int max = 0;
         Stack<LabelWrapper> stack = new Stack<>();
-        Set<LabelWrapper> pushed = new HashSet<>();
+        Map<LabelWrapper, LabelWrapper> pushed = new SmartHashMap<>();
```

Comment:
Why not use `org.jetbrains.kotlin.utils.SmartSet` here?

Reply:
See my PR notes/commit message. SmartSet is only optimized for the 0/1 case and then fallsback to HashSet. From my data analysis, this doesn't make the pushed set a good candidate for SmartSet:

| What               | Amount |
| ------------------ | ------ |
| hashtables created | 4006   |
| max entries        | 175    |
| median entries     | 5      |
| average entries    | 6.10   |
| stdev entries      | 8.28   |
| 90 percentile      | 11     |

It would fallback to a Hashtable in almost all cases, making this a cost instead of a savings

Reply:
It's actually optimized for sizes <= 5, see https://github.com/JetBrains/kotlin/blob/master/core/util.runtime/src/org/jetbrains/kotlin/utils/SmartSet.kt. Maybe you're confusing it with `SmartHashSet` from intellij-core, which is indeed not applicable here?

Reply:
Ah yes, let me give SmartSet a try

Reply:
confirmed that SmartSet removes the performance overhead on object.hashCode from visitMaxs(..) as well. I've applied the change.